gaia/actions.py

```
@@ -0,0 +1,204 @@
+    def __init__(self, hexagon: Hexagon):
+        self.hexagon = hexagon
+        self.base_navigation = 0
+        self.free_gaiaforming = 0
+
+    def validate(self, gamestate, player_id: str) -> Tuple[bool, str]:
+        player = gamestate.players[player_id]
+        game_map = gamestate.game_map
+
+        planet = game_map.get_planet(self.hexagon)
+        if planet is None:
+            return False, "There is not planet on the specified hexagon"
+
+        if isinstance(planet, InhabitedPlanet):
+            return False, "This planet is already occupied"
+
+        navigation_range = self.base_navigation + gamestate.research_board.get_player_navigation_ability(player)
+        planets_in_range = game_map.get_planets_in_range(self.hexagon, navigation_range, only_inhabited=True)
+
+        if not any(planet.faction == player.faction for planet in planets_in_range):
+            return False, "The planet is not in range"
+
+        if planet.planet_type == PlanetType.TRANSDIM:
+            return False, "Cannot build on transdim planets"
+        elif planet.planet_type == PlanetType.GAIA:
+            total_cost = self.cost + Cost(qic=1)
+        else:
+            num_gaiaforms_required = player.get_distance_from_planet_color(planet) - self.free_gaiaforming
+            total_cost = Cost(ore=num_gaiaforms_required * gamestate.research_board.get_player_gaiaforming_cost(player))
+
+        if not player.can_afford(total_cost):
+            return False, "The player cannot afford to place a mine"
+
+        return True, "The player can place a mine at {}".format(str(planet.hex))
+
+    def perform_action(self, gamestate, player_id: str) -> Tuple[bool, str]:
+        player = gamestate.players[player_id]
+        result = gamestate.game_map.inhabit_planet(self.hexagon, player.faction, Building.MINE)
+        return result, ("Successfully built mine"
+                        if result else
+                        "Unable to build mine")
```

Lets raise an error if the build is unsuccessful here. If a move gets past validation and fails that is an issue on our side.